contract/src/main.rs

```diff
@@ -239,58 +345,89 @@ pub extern "C" fn mint() {
 
     // Revert if minting is private and caller is not installer.
     if let MintingMode::Installer = minting_mode {
-        let installer_account = runtime::get_key(INSTALLER)
-            .unwrap_or_revert_with(NFTCoreError::MissingInstallerKey)
-            .into_account()
-            .unwrap_or_revert_with(NFTCoreError::FailedToConvertToAccountHash);
-
-        // Revert if private minting is required and caller is not installer.
-        if caller != installer_account {
-            runtime::revert(NFTCoreError::InvalidMinter)
+        match holder_mode {
+            NFTHolderMode::Accounts => {
+                let installer_account = runtime::get_key(INSTALLER)
+                    .unwrap_or_revert_with(NFTCoreError::MissingInstallerKey)
+                    .into_account()
+                    .unwrap_or_revert_with(NFTCoreError::FailedToConvertToAccountHash);
+
+                // Revert if private minting is required and caller is not installer.
+                if runtime::get_caller() != installer_account {
+                    runtime::revert(NFTCoreError::InvalidMinter)
+                }
+            }
+            NFTHolderMode::Contracts => {
+                let calling_contract = get_calling_contract_hash();
+                let contract_whitelist = get_stored_value_with_user_errors::<Vec<ContractHash>>(
+                    CONTRACT_WHITELIST,
+                    NFTCoreError::MissingWhitelistMode,
+                    NFTCoreError::InvalidWhitelistMode,
+                );
+
+                // Revert if the calling contract is not in the whitelist.
+                if !contract_whitelist.contains(&calling_contract) {
+                    runtime::revert(NFTCoreError::UnlistedContractHash)
+                }
+            }
         }
     }
 
     // The contract's ownership behavior (determined at installation) determines,
     // who owns the NFT we are about to mint.()
-    let ownership_mode = utils::get_ownership_mode().unwrap_or_revert();
+    let ownership_mode = get_ownership_mode().unwrap_or_revert();
     let token_owner_key = {
         match ownership_mode {
-            OwnershipMode::Minter => Key::Account(caller),
+            OwnershipMode::Minter => match holder_mode {
+                NFTHolderMode::Accounts => Key::Account(runtime::get_caller()),
+                NFTHolderMode::Contracts => get_calling_contract_hash().into(),
+            },
```

Comment:
Add `NFTHolderMode::Mixed` which always peeks the immediate caller (which can be either an account or contract)

`NFTHolderMode::Contracts` should fail if the immediate caller is an account

Reply:
Added extra variant, and `Contracts` mode will fail if the immediate caller is an account, added a test to confirm